### PR TITLE
Ebnt 274 api dev

### PR DESCRIPTION
**Pull request by ElmiRakh**

### CODE REVIEW

File: src/ebonite/api/helpers.py [hunk 1]
```diff
@@ -1,4 +1,11 @@
-from pydantic import BaseModel
+from json import loads
+
+import pyjackson as pj
+from pydantic import BaseModel, validator
+
+
+def dumps_pj(pyjackson_obj):
+    return loads(pj.dumps(pyjackson_obj))
```
Comment (mike0sv) [inline src/ebonite/api/helpers.py:8]: pj.serialize

File: src/ebonite/api/helpers.py [hunk 2]
```diff
@@ -19,3 +19,13 @@ class ImageIdValidator(BaseModel):
...
     environment_id: int
+
+
+class BuildableValidator(BaseModel):
+    obj_type: str
+    obj_id: int
+
+    @validator('obj_type')
+    def obj_type_val(cls, v):
+        assert v in ['pipeline', 'model'], 'obj_type should be one of pipeline/model'
```
Comment (mike0sv) [inline src/ebonite/api/helpers.py:30]: technically buildable is anything that can be analyzed by BuidableAnalyzer

File: src/ebonite/api/images.py [hunk 3]
```diff
@@ -4,76 +4,174 @@
...
+            description: Task with given id does not exist
+        """
+        task_id = request.args.get('task_id')
+        TaskIdValidator(task_id=task_id)
+        task = ebonite.meta_repo.get_task_by_id(task_id)
+        if task is not None:
+            return jsonify([pj.serialize(x) for x in ebonite.get_images(task)]), 200
         else:
-            return jsonify({'errormsg': f'Model with id {model_id} does not exist'})
+            return jsonify({'errormsg': f'Task with id {task_id} does not exist'}), 404
```
Comment (mike0sv) [inline src/ebonite/api/images.py:44]: maybe it's will be more readable if you use raise specific exception here and add error handler for it

File: src/ebonite/api/instances.py [hunk 4]
```diff
@@ -38,52 +57,142 @@ def get_instances() -> Tuple[Response, int]:
...
+          404:
+            description: Image with given id does not exist
+        """
+        meta_only = False if not request.args.get('meta_only') else bool(int(request.args.get('meta_only')))
         instance = ebonite.meta_repo.get_instance_by_id(id)
         if instance is None:
             return jsonify({'errormsg': f'Instance with id {id} does not exist'}), 404
-        ebonite.stop_instance(instance)
+        ebonite.delete_instance(instance, meta_only=meta_only)
         return jsonify({}), 204
```
Comment (mike0sv) [inline src/ebonite/api/instances.py:196]: why return empty object in body? either smth like {ok:True} or just empty string will be better i think

File: tests/api/conftest.py [hunk 5]
```diff
@@ -5,12 +5,13 @@
...
-from tests.ext.s3.conftest import s3_artifact, s3server
+from ebonite.core.objects.core import Buildable, Image, Project, RuntimeEnvironment, RuntimeInstance, Task
+from tests.ext.test_s3.conftest import s3_artifact, s3server  # noqa
+from tests.ext.test_sqlalchemy.test_postgres.conftest import postgres_meta, postgres_server  # noqa
 
 
 @pytest.fixture
-def api(tmpdir_factory, postgres_server, postgres_meta, s3server, s3_artifact):
+def api(tmpdir_factory, postgres_server, postgres_meta, s3server, s3_artifact): # noqa
     os.environ['S3_ACCESS_KEY'] = 'eboniteAccessKey'
```
Comment (mike0sv) [inline tests/api/conftest.py:15]: better use constants from s3server

File: tests/api/conftest.py [hunk 6]
```diff
@@ -61,3 +63,58 @@ def create_task_1(client, create_project_1):
 def create_task_2(client, create_project_1):
     rv = client.post('/tasks', data=json.dumps({'name': 'task_2', 'project_id': 1}))
     assert rv.status_code == 201
+
+
+@pytest.fixture
+def project_in_db(postgres_meta, s3_artifact): # noqa
+    proj = Project(name='test_project')
+    proj = postgres_meta.create_project(proj)
+    yield proj
```
Comment (mike0sv) [inline tests/api/conftest.py:72]: you can use return if there is nothing to clear after

File: tests/api/conftest.py [hunk 7]
```diff
@@ -61,3 +63,58 @@ def create_task_1(client, create_project_1):
...
+
+
+@pytest.fixture
+def task_in_db(postgres_meta, project_in_db, s3_artifact): # noqa
```
Comment (mike0sv) [inline tests/api/conftest.py:76]: also why noqa? if you use it, better specify errors you want to be ignored explicitly

File: src/ebonite/api/instances.py (hunk 4 above)
Comment (ElmiRakh) [reply src/ebonite/api/instances.py:196]: 204 response body is always empty

Comment (mike0sv) [reply src/ebonite/api/instances.py:196]: can you just return None, 204 or '', 204?

File: src/ebonite/api/images.py [hunk 8]
```diff
@@ -1,80 +1,176 @@
...
-        body['id'] = id
-        image = UpdateImageBody.from_data(body)
-        try:
-            ebonite.meta_repo.update_image(image)
-            return jsonify({}), 204
-        except NonExistingImageError:
-            return jsonify({'errormsg': f'Project with id {id} does not exist'}), 404
+    # @blueprint.route('/<int:id>', methods=['PATCH'])
+    # def update_image(id: int):
+        # TODO: It's bad. Maybe recreate image when trying to update meta image?
```
Comment (mike0sv) [inline src/ebonite/api/images.py:128]: no, we can update meta before we even build the first version. so lets keep the api as close to client as possible

File: src/ebonite/api/instances.py [hunk 9]
```diff
@@ -38,52 +58,141 @@ def get_instances() -> Tuple[Response, int]:
...
-        body['id'] = id
-        instance = UpdateInstanceBiody.from_data(body)
-        try:
-            ebonite.meta_repo.update_instance(instance)
-            return jsonify({}), 204
-        except NonExistingInstanceError:
-            return jsonify({'errormsg': f'Instance with id {id} does not exist'}), 404
+    # @blueprint.route('/<int:id>', methods=['PATCH'])
+    # def update_instance(id: int) -> Tuple[Response, int]:
+    #     # TODO: Same as with image - to update we need to recreate instance?
```
Comment (mike0sv) [inline src/ebonite/api/instances.py:160]: same

File: src/ebonite/api/pipelines.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+            description: Pipeline with given id does not exist
+          422:
+            description: Incorrect data sent in body
+        """
+        body = request.get_json(force=True)
+        body['id'] = id
+        old_pipeline = ebonite.meta_repo.get_pipeline_by_id(id)
+        if old_pipeline is None:
+            raise ObjectWithIdDoesNotExist('Pipeline', id)
+        # TODO: Should I leave options for this fields to be set through API
```
Comment (mike0sv) [inline src/ebonite/api/pipelines.py:120]: maybe add an update method to EboniteParam (and make EboniteObject subclass of EboniteParam to also use this)
this update method will go through all fields (using pj helpers) and update them (either directly or using nested .update for nested EboniteParam instances)
I did not think this through for every possible case, just so you know :)

**Pull request merged**